pkg/apis/resource/extension.go

```diff
@@ -347,36 +347,97 @@ func (h Handler) getEndpointsFromOutput(ctx context.Context, id object.ID) ([]Ac
 	return endpoints, nil
 }
 
-func (h Handler) getEndpointsFromResources(ctx context.Context, id object.ID) ([]AccessEndpoint, error) {
-	res, err := h.modelClient.ResourceComponents().Query().
-		Where(
-			resourcecomponent.ResourceID(id),
-			resourcecomponent.Mode(types.ResourceComponentModeManaged),
-			resourcecomponent.TypeIn(intercept.TFEndpointsTypes...)).
-		Select(
-			resourcecomponent.FieldConnectorID,
-			resourcecomponent.FieldType,
-			resourcecomponent.FieldName,
-			resourcecomponent.FieldStatus,
-		).
-		All(ctx)
+func (h Handler) getEndpointsFromResources(ctx context.Context, id object.ID, wait bool) ([]AccessEndpoint, error) {
+	sr, err := h.getLatestRevision(ctx, id)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(res) == 0 {
+	if sr == nil {
 		return nil, nil
 	}
 
-	var endpoints []AccessEndpoint
+	if !status.ResourceRevisionStatusReady.IsTrue(sr) {
+		return nil, nil
+	}
 
-	for _, v := range res {
-		for _, eps := range v.Status.ResourceEndpoints {
-			endpoints = append(endpoints, AccessEndpoint{
-				Name:      filepath.Join(eps.EndpointType, filepath.Base(v.Name)),
-				Endpoints: eps.Endpoints,
-			})
+	getEndpoint := func(ctx context.Context, id object.ID) ([]AccessEndpoint, error) {
+		res, err := h.modelClient.ResourceComponents().Query().
+			Where(
+				resourcecomponent.ResourceID(id),
+				resourcecomponent.Mode(types.ResourceComponentModeManaged),
+				resourcecomponent.TypeIn(intercept.TFEndpointsTypes...)).
+			Select(
+				resourcecomponent.FieldConnectorID,
+				resourcecomponent.FieldType,
+				resourcecomponent.FieldName,
+				resourcecomponent.FieldStatus,
+			).
+			All(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(res) == 0 {
+			return nil, nil
+		}
+
+		var endpoints []AccessEndpoint
+
+		for _, v := range res {
+			for _, eps := range v.Status.ResourceEndpoints {
+				endpoints = append(endpoints, AccessEndpoint{
+					Name:      filepath.Join(eps.EndpointType, filepath.Base(v.Name)),
+					Endpoints: eps.Endpoints,
+				})
+			}
 		}
+
+		return endpoints, nil
+	}
+
+	if !wait {
+		return getEndpoint(ctx, id)
+	}
+
+	// Wait for the resource component status's ResourceEndpoints update.
+	expr, err := cron.ParseCronExpr(
+		settings.ResourceComponentStatusSyncCronExpr.ShouldValue(ctx, h.modelClient),
+		false)
+	if err != nil {
+		return nil, err
+	}
+
+	timeout := time.Until(
+		expr.
+			Next(time.Now()).
+			Add(10 * time.Second))
+
+	if timeout > 180*time.Second {
+		timeout = 180 * time.Second
+	}
+
+	waitCtx, cancel := context.WithTimeout(ctx, timeout)
```

Comment:
What is the user experience for user if wait timeout?

Reply:
Need refresh to load the access endpoint. In our framework, updating events are generated through changes in database data. It is challenging to support both high availability (HA) and custom triggering simultaneously. Modifying the revision to trigger data changes when updating the component endpoint in a scheduled task is not appropriate. Create another task to support trigger customized event https://github.com/seal-io/walrus/issues/1682